### PR TITLE
allow Pagination in Collaborator List

### PR DESCRIPTION
**Pull request by ibrahimjaved12**
# What are the relevant tickets?
Closes https://github.com/mitodl/hq/issues/1162

# Description (What does it do?)
This PR enables pagination in Collaborator List.

# How can this be tested?
1. Checkout to this branch
2. `docker-compose up`
3. Make some users from django admin if you don't have already (we are looking for 10+ users) (also make sure they are not superusers)
4. Go to Collaborators of any site, eg. http://localhost:8043/sites/ocw-www/collaborators/
5. Add collaborators, at 11 collaborators the Pagination component should appear at the bottom of the page. At 21 collaborators, we should have both (Next) and (Prev) icons on the 2nd page.
6. Make sure to test the edge cases; Add/Delete/Modify changes work. And they should be reflected immediately in the collaborators list too. Try adding a new collaborator, and then modifying it, and then deleting it, to make sure the API calls work as expected. Check inspect element / refresh the page to ensure there are no errors.
7. You can also see from Django admin, under users, if a user is a collaborator, then he should have either Editors permissions or Admins permissions under groups. And when you modify it / delete collaborator, the changes should be reflected in the admin panel after refreshing the page.

### CODE REVIEW

**Review by HussainTaj-arbisoft — CHANGES_REQUESTED** [review]
Good work, the functionality works correctly. I have a few suggestions.

Is it possible for us to make this simpler? In particular, I'm talking about the state transformations we're doing. If I understand correctly, we don't have a collaborator details API, right? We get the response from the listing API and transform it into objects in the state (namely `collaborators` and `websiteCollaboratorDetails`). This is to simulate a separate listing and details endpoints. It makes sense to me to follow existing patterns to make things easier to understand.

I'd like to ask if it will help not to transform the API response into `collaborators` and `websiteCollaboratorDetails` and use only `collaborator`. With some well-placed selectors, it might be possible to avoid some extra steps of data transformations.

I'm not sure if I explained it well, do let me know if you have any questions.

File: static/js/query-configs/websites.ts [hunk 1]
```diff
@@ -261,14 +225,6 @@ export const editWebsiteCollaboratorMutation = (
...
       })
       .toString(),
-    optimisticUpdate: {
-      collaborators: evolve({
-        [websiteName]: compose(
-          alterRole,
-          (value: WebsiteCollaborator[]) => value || [],
-        ),
-      }),
-    },
```
Comment (ibrahimjaved12) [inline static/js/query-configs/websites.ts:271]: Don't need optimisticUpdate now. The state is already updated by fetch() call.

**Review by HussainTaj-arbisoft — COMMENTED** [review]
The functionality works well. And the new code is a lot simpler. Thank you!

I just have a few questions.

File: static/js/components/SiteCollaboratorList.test.tsx [hunk 2]
```diff
@@ -20,20 +20,54 @@ import WebsiteContext from "../context/Website"
...
+      previous: null,
+    }
+    const collaboratorListingState = {
+      [collaboratorListingKey(listingParams)]: {
+        ...apiResponse,
+        results: apiResponse.results.map(
+          (collaborator) => collaborator.user_id,
+        ),
+      },
+    }
```
[the remark this answers is not in the document]
Comment (ibrahimjaved12) [reply static/js/components/SiteCollaboratorList.test.tsx:61]: Yes. The `...apiResponse` should have the results, and we do not map by `collaborator.user_id` anymore

File: static/js/selectors/websites.ts [hunk 3]
```diff
@@ -46,19 +48,25 @@ export const getWebsiteListingCursor = createSelector(
...
-  (collaborators) =>
-    memoize((name: string, username: string) =>
-      find(propEq("username", username), collaborators[name]),
+  (listing) =>
+    memoize(
+      (
+        listingParams: CollaboratorListingParams,
+      ): WebsiteCollaboratorListSelection => {
+        const response = listing[collaboratorListingKey(listingParams)] ?? {}
+        return { ...response }
```
[the remark this answers is not in the document]
Comment (ibrahimjaved12) [reply static/js/selectors/websites.ts:63]: Should work the same way

**Comment by ibrahimjaved12** [comment]
Thanks for these suggestions. I've updated the code. Over to you again @HussainTaj-arbisoft
